servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/OptionalSslNegotiator.java

```diff
@@ -0,0 +1,125 @@
+                                          @Nullable final LateConnectionAcceptor lateConnectionAcceptor) {
+        final ReadOnlyHttpServerConfig roConfig = config.asReadOnly();
+        final ReadOnlyTcpServerConfig roTcpConfig = roConfig.tcpConfig();
+
+        final BiFunction<Channel, ConnectionObserver, Single<NettyConnectionContext>> channelInit =
+                (channel, connectionObserver) ->
+                        new OptionalSslChannelSingle(channel, OptionalSslChannelSingle.NoopChannelInitializer.INSTANCE)
+                                .flatMap(isTls -> {
+                                    if (isTls) {
+                                        if (roTcpConfig.isAlpnConfigured()) {
+                                            return DeferredServerChannelBinder.alpnInitChannel(listenAddress, channel,
+                                                    roConfig, executionContext,
+                                                    service, drainRequestPayloadBody, connectionObserver);
+                                        } else if (roTcpConfig.sniMapping() != null) {
+                                            return DeferredServerChannelBinder.sniInitChannel(listenAddress, channel,
+                                                    roConfig, executionContext,
+                                                    service, drainRequestPayloadBody, connectionObserver);
+                                        } else if (roConfig.isH2PriorKnowledge()) {
+                                            return H2ServerParentConnectionContext.initChannel(listenAddress, channel,
+                                                    executionContext, roConfig,
+                                                    new TcpServerChannelInitializer(roTcpConfig, connectionObserver,
+                                                            executionContext), service,
+                                                    drainRequestPayloadBody, connectionObserver);
+                                        } else {
+                                            return NettyHttpServer.initChannel(channel, executionContext, roConfig,
+                                                    new TcpServerChannelInitializer(roTcpConfig, connectionObserver,
+                                                            executionContext), service,
+                                                    drainRequestPayloadBody, connectionObserver);
+                                        }
+                                    }
+
+                                    // TODO: need to duplicate all but SSL config
+                                    final HttpServerConfig newConfig = new HttpServerConfig();
+                                    if (roConfig.h1Config() != null && roConfig.h2Config() != null) {
+                                        newConfig.httpConfig().protocols(roConfig.h1Config(), roConfig.h2Config());
```

Comment:
This won't work for plaintext, should we fallback to HTTP/1.1?
We can consider adding a log statement in `DefaultHttpServerBuilder` (either inside `listenForService` or `doBind`) to let users know that we can fallback to HTTP/1.1 for plaintext path if they have both `acceptInsecureConnections()` and 2 protocols configured.

Reply:
We can consider something like this, but the TODO above is a bit wider. Do you know how we could duplicate a server config easily modulo the ssl bits? Or somehow duplicate the builder and then convert to readonly.